hasprefix.go

```diff
@@ -0,0 +1,40 @@
+package vfs
+
+import (
+	"os"
+	"path/filepath"
+)
+
+// A Stater implements Stat.
+type Stater interface {
+	Stat(string) (os.FileInfo, error)
+}
+
+// HasPrefix returns true if p or any parent of p is the same file as prefix.
+// prefix must exist, but p may not. It is an expensive but accurate alternative
+// to the deprecated filepath.HasPrefix.
+func HasPrefix(fs Stater, p, prefix string) (bool, error) {
+	prefixFI, err := fs.Stat(prefix)
+	if err != nil {
+		return false, err
+	}
+	for {
+		fi, err := fs.Stat(p)
+		switch {
+		case err == nil:
+			if os.SameFile(fi, prefixFI) {
+				return true, nil
+			}
+		case os.IsNotExist(err):
+			// Do nothing and skip ahead to trying p's parent directory.
+		default:
+			return false, err
+		}
+		parentDir := filepath.Dir(p)
```

Comment:
There needs to be some thought about how you want to deal with symlinks etc. For example, if `p` denotes a symlink *into* `prefix` (say `prefix == "/home/mero"` and `p` is `/tmp/foo`, pointing at `/home/mero/foo`), should this function return `true`, or `false`? And on the other hand, if `p` is a symlink *inside* `prefix`, pointing *out* of it (say `prefix == /home/mero` and `p` is `/home/mero/foo` pointing at `/tmp/foo`)? It really depends on the semantic of what you want to achieve.

Just as food for thought: It would be possible to use `Open` in combination with [Openat](https://godoc.org/golang.org/x/sys/unix#Openat) to walk the actual inodes. That should make this totally, 100% referring to the on-disk structures, respecting all permissions and symlinks and the like. Would obviously only work on posix operating systems. I *think* that mainly excludes Plan 9 currently? Dunno.

Again, depends on the semantics that you want to achieve.

Reply:
That's really insightful. I've updated the comment:

```go
// Contains returns true if p is reachable by traversing through prefix. prefix
// must exist, but p may not. It is an expensive but accurate alternative to the
// deprecated filepath.HasPrefix.
```

I _think_ this is enough for now, but you're right that there's a lot of subtlety here.